### ISSUE TITLE
Allow export_studies to export to an existing directory

### ISSUE THREAD

**Issue by lparsons** [Princeton-LSI-ResearchComputing/tracebase#808]
### Inspiration

There are various sceanrios where one might want to run `export_studies` multiple times with different parameters and add new files to an export directory. The current script requires that the `outdir` be a new directory that does not exist at the time the script is run. This makes it difficult to incrementally add files.

### Description

Allow the script to export to an already existing directory.

### Alternatives

None

### Comment

None

-----

### Assumptions

1. List of assumptions that the code will not explicitly address/check
2. E.g. We will assume input is correct (explaining why there is no validation)

### Limitations

1. A list of things this work will specifically not do
2. E.g. This feature will only handle the most frequent use case X

### Affected Components

- change: File path or DB table ...
- add: Environment variable or server setting
- delete: External executable or cron job

### Requirements

- [ ] 1. List of numbered conditions to be met for the feature
- [ ] 2. E.g. Every column/row must display a value, i.e. cannot be empty
- [ ] 3. Numbers for reference & checkboxes for progress tracking

#### Interface Change description

None provided

#### Code Change Description

None provided

#### Tests

- [ ] 1. A description of at least one test for each requirement above.
- [ ] 2. E.g. Test for req 2 that there's an exception when display value is ''
- [ ] 3. Numbers for reference & checkboxes for progress tracking

**Comment by lparsons** [comment]
@hepcat72 Can you take a quick look at the proposed design for `export_studies`. Summary is that I plan to add a `--force` flag to allow the script to overwrite existing files. I also plan to create the files in TMPDIR and move them to the OUTDIR when complete to avoid having partial files there for an extended period of time.

**Comment by hepcat72** [comment]
I think you should add:
- assumption: **if a file exists, it is complete** _(because you're not checking the content of existing files [when `--force` is not supplied])_
- limitation: this does not check or remove old/stale files related to deleted/non-existant study data

WRT the design...
```
if OUTDIR doesn't exist, error
if STUDYDIR doesn't exist, create
if DATAFILE exists and  NOT FORCE, error
```
If you want to error out when `OUTDIR` doesn't exist, does that mean you'll be removing `os.mkdir(self.outdir)` and expect the user to create it before running the script?  Why not `create` if it didn't exist?

In fact, I would probably skip if `DATAFILE` exists and `NOT FORCE`.  Force would just overwrite.  In short:
```
if OUTDIR doesn't exist, create
if STUDYDIR doesn't exist, create
if DATAFILE doesn't exist or FORCE, create(/overwrite)
```
In my mind, the only reason to error would be to head off unexpected outcomes/data.  And the reason for the error on an existing directory was to prevent unexpected overwriting of existing data.  If this is careful to never overwrite existing data (unless explicitly indicated by the user), it solves that problem.  The new problem it introduces is unexpected old/stale data mixed in with the expected output, so the only error I might expect from this would be to call out files that were unexpected (i.e. not associated with data in the current DB due to deletions/usage of a different DB, or inaccurate due to changes in the DB).

And if you plan to create a temp dir, it might be a modicum complex to implement the existing file logic?  How do you plan to handle that?  Per file?  Instead of a tmp dir, you could use a `.tmp` extension...

I'm not saying you should make these changes. It's just my opinion/preference.  I.e. you can proceed with your design without making any changes to your plan as far as I'm concerned.

**Comment by hepcat72** [comment]
BTW, I have a PR I'm about to submit for the export exceptions. I don't think it would conflict with this effort, but it does change 1 line in the export code and some of the tests.

**Comment by lparsons** [comment]
> I think you should add:
> 
> * assumption: **if a file exists, it is complete** _(because you're not checking the content of existing files [when `--force` is not supplied])_

I disagree. This is not assumed. If a file exists and `--force` is not supplied it will not be overwritten, regardless of contents.

> * limitation: this does not check or remove old/stale files related to deleted/non-existant study data

This seems like it's already covered to me by "Extra files/directories in outdir will not be removed"

Good catches on the some of the logic points. I tweaked things a bit to make them a bit cleaner. I do like the `.tmp` file idea. I  think I'll implement that instead of the tempfile since tempfiles sometimes exist on different mounts and thus can't be easily "moved".

**Comment by hepcat72** [comment]
> > I think you should add:
> > 
> > * assumption: **if a file exists, it is complete** _(because you're not checking the content of existing files [when `--force` is not supplied])_
> 
> I disagree. This is not assumed. If a file exists and `--force` is not supplied it will not be overwritten, regardless of contents.

I should have qualified that this applies only to the case of not supplying `--force`.  It is an assumption, but only in that case.  The fact that it previously required the directory not to exist had prevented that from being an assumption.

> > * limitation: this does not check or remove old/stale files related to deleted/non-existant study data
> 
> This seems like it's already covered to me by "Extra files/directories in outdir will not be removed"

Fair, though not referring to stale/old data leads the reader to think it only applies to notes or files unrelated to previous exports, and that's the part that seems vulnerable to me.

> Good catches on the some of the logic points. I tweaked things a bit to make them a bit cleaner. I do like the `.tmp` file idea. I think I'll implement that instead of the tempfile since tempfiles sometimes exist on different mounts and thus can't be easily "moved".

👍🏻
